### PR TITLE
Support JSON-LD document with context array and undefined terms

### PR DESCRIPTION
**Pull request by zhenwenc**
This PR enhances the JSON-LD compression and CBOR-LD decompression to support a wider range of use cases.

## Support parsing `@context` array recursively

The previous implementation only supports the top-level `@context` in the JSON-LD document to be an array, but this use case could happen in remote contexts too (e.g. https://www.w3.org/2018/credentials/examples/v1):

```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://www.w3.org/2018/credentials/examples/v1"
  ],
  "type": ["VerifiableCredential"],
  "issuer": {
    "id": "did:key:z6Mknra61Bne37pVAnZD6QxXbsUL5Nu9TRPKmSN4VWSAu7D6",
    "name": "Term defined in the example context, but couldn't be detected correctly"
  },
  "credentialSubject": {
    "id": "did:key:z6MktojtXugNxXuhypwYxjjZM5Wk1FqBnwwyJzhrnKsDpQHX"
  },
  "proof": { ... }
}
```

## Support undefined terms in the JSON-LD/CBOR-LD document

There are cases that a JSON-LD document contains terms that are not defined in any of the declared contexts, based on the JSON-LD specification, these terms will be expanded with the `@vocab` [default vocabulary](https://www.w3.org/TR/json-ld11/#default-vocabulary), which wasn't supported by this library.

Instead of failing the encode/decode operations will `ERR_UNKNOWN_CBORLD_TERM_ID` error, we allow uses to optionally skip compressing the unknown terms by preserving the original term key, which can be controlled by the `allowUndefinedTerms: boolean` option in both `encode` and `decode` functions (defaults to `false`).

This will enable us to encode JSON-LD documents with compression like this example, where the "proof.proofValue" term hasn't been defined: https://github.com/w3c/vc-test-suite/blob/cfad6f390316529e32ef0a42d1330e16ba42968c/test/vc-data-model-1.0/input/example-015-zkp.jsonld

Erro with `allowUndefinedTerms: false`:
```
Error: ERR_UNKNOWN_CBORLD_TERM: Unknown term 'proofValue' was detected in the JSON-LD input.
 at new CborldError (lib/CborldError.js:9:19)
 at Compressor._getEntries (lib/Compressor.js:134:15)
 at Compressor._transform (lib/Transformer.js:56:30)
 at processTicksAndRejections (internal/process/task_queues.js:97:5)
 at async Compressor._transformObject (lib/Compressor.js:191:5)
 at async Compressor._transform (lib/Transformer.js:122:9)
 at async Compressor._createTransformMaps (lib/Compressor.js:90:7)
 at async Compressor.compress (lib/Compressor.js:70:27)
 at async encode (lib/encode.js:53:14)
```

When encoding with `allowUndefinedTerms: true`, the following transform map will be produced after compression:
```
Diagnostic CBOR-LD compression transform map(s):
Map {
  1 => [
    ContextEncoder {
      context: 'https://www.w3.org/2018/credentials/v1',
      appContextMap: Map {}
    },
    ContextEncoder {
      context: 'https://www.w3.org/2018/credentials/examples/v1',
      appContextMap: Map {}
    }
  ],
  114 => Map {
    'proofValue' => 'DgYdYMUYHURJLD7xdnWRinqWCEY5u5fK...j915Lt3hMzLHoPiPQ9sSVfRrs1D',
    116 => 'ex:AnonCredPresentationProofv1'
  },
  116 => VocabTermEncoder { termId: 110 },
  542 => Map {
    1 => [
      ContextEncoder {
        context: 'https://www.w3.org/2018/credentials/v1',
        appContextMap: Map {}
      },
      ContextEncoder {
        context: 'https://www.w3.org/2018/credentials/examples/v1',
        appContextMap: Map {}
      }
    ],
    544 => Map {
      112 => 'did:example:cdf:35LB7w9ueWbagPL94T9bMLtyXDj9pX5o',
      116 => 'did:example:schema:22KpkXgecryx9k7N6XN1QoN3gXwBkSU8SfyyYQG'
    },
    548 => Map { 500 => 'College of Engineering', 502 => 'BachelorDegree' },
    554 => XsdDateTimeEncoder {
      value: '2010-01-01T19:23:24Z',
      parsed: 1262373804000
    },
    558 => 'did:example:Wz4eUg7SetGfaUVCn8U9d62oDYrUJLuUtcy619',
    114 => Map {
      518 => 'mu6fg24MfJPU1HvSXsf3ybzKARib4WxG...RSce53M6UwQCxYshCuS3d2h',
      522 => 'cg7wLNSi48K5qNyAVMwdYqVHSMv1Ur8i...Fg2ZvWF6zGvcSAsym2sgSk737',
      116 => 'ex:AnonCredDerivedCredentialv1'
    },
    117 => [
      VocabTermEncoder { termId: 108 },
      VocabTermEncoder { termId: 486 }
    ]
  }
}
```

Note that the original key of the undefined term `proofValue` had been preserved.

## Introduced integration tests with JSON-LD documents in vc-test-suite

To ensure we cover the edge cases of Verifiable Credential documents.

### CODE REVIEW

File: lib/Decompressor.js [hunk 1]
```diff
@@ -36,9 +36,12 @@ export class Decompressor extends Transformer {
    * @param {Map} [options.appContextMap] - A map of JSON-LD Context URLs and
    *   their encoded CBOR-LD values (must be values greater than 32767
    *   (0x7FFF)).
+   * @param {boolean} [options.allowUndefinedTerms] - Allow the JSON-LD document
+   *   to contain terms that are not defined in the context. The original values
+   *   of terms that cannot be resolved from the context map will be preserved.
```
Comment (aliXsed) [inline lib/Decompressor.js:41]: Perhaps, good to mention here that: `The key type is expected to be a string for undefined terms`

**Comment by aliXsed** [comment]
It could be good to split this PR to 2 or 3 separate PRs. So as soon as one is approved you can merge that individually. Just a suggestion though.

**Comment by zhenwenc** [comment]
This PR had been split into:

- https://github.com/mattrglobal/digitalbazaar-cborld/pull/2
- https://github.com/mattrglobal/digitalbazaar-cborld/pull/3
